---
author: Rishabh Narula (narula.r)
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    dev: 'svg'
  md_document:
    variant: gfm
bibliography: bibliography.ris
---


```{bash, error=T}

# Getting file from using hapmap1

plink --file hapmap1 

# Make binary PED file

plink --file hapmap1 --make-bed --out hapmap1 

plink --file hapmap1 --make-bed --mind 0.05 --out highgeno 

plink --bfile hapmap1 

plink --bfile hapmap1 --missing --out miss_stat 

#more miss_stat.lmiss 

plink --bfile hapmap1 --chr 1 --out res1 --missing 

plink --bfile hapmap1 --chr 2 --out res2 --missing 

plink --bfile hapmap1 --freq --out freq_stat 

plink --bfile hapmap1 --freq --within pop.phe --out freq_stat 

#more freq_stat.frq.strat 

plink --bfile hapmap1 --snp rs1891905 --freq --within pop.phe --out snp1_frq_stat 

plink --bfile hapmap1 --assoc --out as1 

sort --key=7 -nr as1.assoc | head 

plink --bfile hapmap1 --assoc --adjust --out as2 

#more as2.assoc.adjusted 

plink --bfile hapmap1 --pheno pop.phe --assoc --adjust --out as3 

plink --bfile hapmap1 --model --snp rs2222162 --out mod1

plink --bfile hapmap1 --model --cell 0 --snp rs2222162 --out mod2 

plink --bfile hapmap1 --cluster --mc 2 --ppc 0.05 --out str1 

#more str1.cluster1 

plink --bfile hapmap1 --mh --within str1.cluster2 --adjust --out aac1 

#more aac1.cmh.adjusted 

plink --bfile hapmap1 --cluster --cc --ppc 0.01 --out version2 

plink --bfile hapmap1 --mh --within version2.cluster2 --adjust --out aac2 

plink --bfile hapmap1 --cluster --K 2 --out version3 

plink --bfile hapmap1 --mh --within pop.phe --adjust --out aac3 

plink --bfile hapmap1 --cluster --matrix --out ibd_view 




plink --bfile hapmap1 --assoc --pheno qt.phe --out quant1 

plink --bfile hapmap1 --assoc --pheno qt.phe --perm --within str1.cluster2 --out quant2 

plink --bfile hapmap1 --assoc --pheno qt.phe --mperm 1000 --within str1.cluster2 --out quant3 

plink --bfile hapmap1 --pheno qt.phe --gxe --covar pop.phe --snp rs2222162 --out quant3

plink --bfile hapmap1 --snp rs2222162 --recodeAD --out rec_snp1 

```


```{R, echo=TRUE, eval=TRUE}

m <- as.matrix(read.table("ibd_view.mibs")) 

mds <- cmdscale(as.dist(1-m))

k <- c( rep("green",45) , rep("blue",44) ) 

plot(mds,pch=20,col=k) 

```
where each row is a single SNP association result. The fields are: Chromosome, SNP identifier, Code for allele 1 (the minor, rare allele based on the entire sample frequencies), The frequency of this variant in cases, The frequency of this variant in controls, Code for the other allele, The chi-squared statistic for this test (1 df), The asymptotic significance value for this test, The odds ratio for this test
If a test is not defined (for example, if the variant is monomorphic but was not excluded by the filters)
then values of NA for not applicable will be given (as these are read by the package R to indicate missing
data, which is convenient if using R to analyse the set of results). 

```{R, echo=TRUE, eval=TRUE}

knitr::kable(head(m)) 

d <- read.table("rec_snp1.raw" , header=T) 

summary(glm(PHENOTYPE-1 ~ rs2222162_1, data=d, family="binomial")) 

```

This particular recode feature codes genotypes as additive (0,1,2) and dominance (0,1,0) components, in a file called rec_snp1.recode.raw. We can then load this file into our statistics package and easily perform other analyses: for example, to repeat the main analysis as a simple logistic regression using the R package (not controlling for clusters):

```{R, echo=TRUE, eval=TRUE}

knitr::kable(head(d))

```
[@chow]

## Refrences

1 : http://hpc.ilri.cgiar.org/beca/training/data_mgt_2017/BackgroundMaterial/PlinkTutorial.pdf
