---
title: "MetaGenomics"
output: github_document
---

```{r, echo=TRUE}

library(metagenomeSeq)

library(biomformat)

biom_file <- system.file("extdata", "min_sparse_otu_table.biom", package = "biomformat")

b <- read_biom(biom_file)

biom2MRexperiment(b)

data(mouseData)

b <- MRexperiment2biom(mouseData)

write_biom(b, biom_file = "otu_table.biom")


```



```{r, echo=TRUE}

dataDirectory <- system.file("extdata", package = "metagenomeSeq")

lung = loadMeta(file.path(dataDirectory, "CHK_NAME.otus.count.csv"))

dim(lung$counts)

taxa = read.delim(file.path(dataDirectory, "CHK_otus.taxonomy.csv"),stringsAsFactors = FALSE)

clin = loadPhenoData(file.path(dataDirectory, "CHK_clinical.csv"), tran = TRUE)

ord = match(colnames(lung$counts), rownames(clin))

clin = clin[ord, ]

head(clin[1:2, ])

```


```{r, echo=TRUE}

phenotypeData = AnnotatedDataFrame(clin)

phenotypeData

OTUdata = AnnotatedDataFrame(taxa)

OTUdata

obj = newMRexperiment(lung$counts,phenoData=phenotypeData,featureData=OTUdata)

obj

data(lungData)

lungData

data(mouseData)

mouseData

phenoData(obj)

head(pData(obj), 3)

featureData(obj)

head(fData(obj)[, -c(2, 10)], 3)

head(MRcounts(obj[, 1:2]))

featuresToKeep = which(rowSums(obj) >= 100)

samplesToKeep = which(pData(obj)$SmokingStatus == "Smoker")

obj_smokers = obj[featuresToKeep, samplesToKeep]

obj_smokers

head(pData(obj_smokers), 3)

head(normFactors(obj))

normFactors(obj) <- rnorm(ncol(obj))

head(normFactors(obj))

head(libSize(obj))

libSize(obj) <- rnorm(ncol(obj))

head(libSize(obj))

data(mouseData)

filterData(mouseData, present = 10, depth = 1000)



```

```{r, echo=TRUE}

data(mouseData)

newobj = mergeMRexperiments(mouseData, mouseData)

newobj

data(lungData)

p = cumNormStatFast(lungData)

lungData = cumNorm(lungData, p = p)

condition = mouseData$diet

mouseData = wrenchNorm(mouseData, condition = condition)

mat = MRcounts(lungData, norm = TRUE, log = TRUE)[1:5, 1:5]

exportMat(mat, file = file.path(dataDirectory, "tmp.tsv"))

exportStats(lungData[, 1:5], file = file.path(dataDirectory, "tmp.tsv"))

head(read.csv(file = file.path(dataDirectory, "tmp.tsv"), sep = "\t"))


```

```{r, echo=TRUE}

data(lungData)

lungData = lungData[, -which(is.na(pData(lungData)$SmokingStatus))]

lungData = filterData(lungData, present = 30, depth = 1)

lungData <- cumNorm(lungData, p = 0.5)

pd <- pData(lungData)

mod <- model.matrix(~1 + SmokingStatus, data = pd)

lungres1 = fitFeatureModel(lungData, mod)

head(MRcoefs(lungres1))

data(lungData)

controls = grep("Extraction.Control", pData(lungData)$SampleType)

lungTrim = lungData[, -controls]

rareFeatures = which(rowSums(MRcounts(lungTrim) > 0) < 10)

lungTrim = lungTrim[-rareFeatures, ]

lungp = cumNormStat(lungTrim, pFlag = TRUE, main = "Trimmed lung data")

lungTrim = cumNorm(lungTrim, p = lungp)

smokingStatus = pData(lungTrim)$SmokingStatus

bodySite = pData(lungTrim)$SampleType

normFactor = normFactors(lungTrim)

normFactor = log2(normFactor/median(normFactor) + 1)

mod = model.matrix(~smokingStatus + bodySite + normFactor)

settings = zigControl(maxit = 10, verbose = TRUE)

fit = fitZig(obj = lungTrim, mod = mod, useCSSoffset = FALSE, control = settings)

```


```{r, echo=TRUE}

settings = zigControl(maxit = 1, verbose = FALSE)

mod = model.matrix(~bodySite)

colnames(mod) = levels(bodySite)


res = fitZig(obj = lungTrim, mod = mod, control = settings)

zigFit = slot(res, "fit")

finalMod = slot(res, "fit")$design

contrast.matrix = makeContrasts(BAL.A - BAL.B, OW - PSB, levels = finalMod)

fit2 = contrasts.fit(zigFit, contrast.matrix)

fit2 = eBayes(fit2)

topTable(fit2)

taxa = sapply(strsplit(as.character(fData(lungTrim)$taxa), split = ";"),
  function(i) {
    i[length(i)]
  })

head(MRcoefs(fit, taxa = taxa, coef = 2))


coeffOfInterest = 2

res = fitLogNormal(obj = lungTrim, mod = mod, useCSSoffset = FALSE, B = 10, coef = coeffOfInterest)

adjustedPvalues = p.adjust(res$p, method = "fdr")

foldChange = abs(res$fit$coef[, coeffOfInterest])

sigList = which(adjustedPvalues <= 0.05)

sigList = sigList[order(foldChange[sigList])]

head(taxa[sigList])

classes = pData(mouseData)$diet

res = fitPA(mouseData[1:5, ], cl = classes)

head(res)

classes = pData(mouseData)$diet

res = fitDO(mouseData[1:100, ], cl = classes, norm = FALSE, log = FALSE)

head(res)

```

```{r, echo=TRUE}

cors = correlationTest(mouseData[55:60, ], norm = FALSE, log = FALSE)

head(cors)

cl = pData(mouseData)[["diet"]]

uniqueFeatures(mouseData, cl, nsamples = 10, nreads = 100)

obj = aggTax(mouseData, lvl = "phylum", out = "matrix")

head(obj[1:5, 1:5])

obj = aggSamp(mouseData, fct = "mouseID", out = "matrix")

head(obj[1:5, 1:5])


```

```{r, echo=TRUE}

trials = pData(mouseData)$diet

heatmapColColors = brewer.pal(12, "Set3")[as.integer(factor(trials))]

heatmapCols = colorRampPalette(brewer.pal(9, "RdBu"))(50)

plotMRheatmap(obj = mouseData, n = 200, cexRow = 0.4, cexCol = 0.4, trace = "none", col = heatmapCols, ColSideColors = heatmapColColors)
# plotCorr
plotCorr(obj = mouseData, n = 200, cexRow = 0.25, cexCol = 0.25, trace = "none", dendrogram = "none", col = heatmapCols)

cl = factor(pData(mouseData)$diet)

# plotOrd - can load vegan and set distfun = vegdist and use
# dist.method='bray'
plotOrd(mouseData, tran = TRUE, usePCA = FALSE, useDist = TRUE, bg = cl, pch = 21)

# plotRare
res = plotRare(mouseData, cl = cl, pch = 21, bg = cl)

# Linear fits for plotRare / legend
tmp = lapply(levels(cl), function(lv) lm(res[, "ident"] ~ res[, "libSize"] - 1, subset = cl == lv))

for (i in 1:length(levels(cl))) {
  abline(tmp[[i]], col = i)
}

legend("topleft", c("Diet 1", "Diet 2"), text.col = c(1, 2), box.col = NA)

head(MRtable(fit, coef = 2, taxa = 1:length(fData(lungTrim)$taxa)))

patients = sapply(strsplit(rownames(pData(lungTrim)), split = "_"),
  function(i) {
    i[3]
  })

pData(lungTrim)$patients = patients

classIndex = list(smoker = which(pData(lungTrim)$SmokingStatus == "Smoker"))

classIndex$nonsmoker = which(pData(lungTrim)$SmokingStatus == "NonSmoker")

otu = 779

# plotOTU
plotOTU(lungTrim, otu = otu, classIndex, main = "Neisseria meningitidis")

# Now multiple OTUs annotated similarly
x = fData(lungTrim)$taxa[otu]
otulist = grep(x, fData(lungTrim)$taxa)

# plotGenus
plotGenus(lungTrim, otulist, classIndex, labs = FALSE, main = "Neisseria meningitidis")

lablist <- c("S", "NS")

axis(1, at = seq(1, 6, by = 1), labels = rep(lablist, times = 3))

classIndex = list(Western = which(pData(mouseData)$diet == "Western"))

classIndex$BK = which(pData(mouseData)$diet == "BK")

otuIndex = 8770

# par(mfrow=c(1,2))
dates = pData(mouseData)$date

plotFeature(mouseData, norm = FALSE, log = FALSE, otuIndex, classIndex, col = dates, sortby = dates, ylab = "Raw reads")


```

```{r, echo=TRUE}

citation("metagenomeSeq")

sessionInfo()









```



