---
author: Rishabh Narula (narula.r)
output:
  md_document:
    variant: markdown_github
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    dev: 'svg'
---

## Part 1

Installing Libraries and essential packages

```{r setup, include=FALSE}

#install.packages("devtools")
library(devtools)
install_github(repo = "PoisonAlien/maftools")
library(maftools)

```

## Part 6
## Reading and Summarizing maf files

We are using system.file() function to read the github files, from maftools package in github. MAF fikes can be very long, so its recommended to provide annotation with samples first. Then we are also using read.maf() function to create MAF object called laml.

```{r, echo=TRUE}

#path to TCGA LAML MAF file
laml.maf = system.file('extdata', 'tcga_laml.maf.gz', package = 'maftools') 
#clinical information containing survival information and histology. This is optional
laml.clin = system.file('extdata', 'tcga_laml_annot.tsv', package = 'maftools') 

laml = read.maf(maf = laml.maf, clinicalData = laml.clin)

```

## MAF Object

Exploring the laml object with different functions using getSampleSummary(),
getClinicalData(), getFields(), and writing that into MafSummary file.

```{r, echo=TRUE}

#Typing laml shows basic summary of MAF file.
laml

#Shows sample summry.
getSampleSummary(laml)

#Shows gene summary.
getGeneSummary(laml)

#shows clinical data associated with samples
getClinicalData(laml)

#Shows all fields in MAF
getFields(laml)

#Writes maf summary to an output file with basename laml.
write.mafSummary(maf = laml, basename = 'laml')


```

## Visualization

Plotting MafSummary object using plotSummary function. It includes Variant Classification, Variant type, SNV class, Variant per sample, top 10 mutates genes.

```{r, echo=TRUE}

plotmafSummary(maf = laml, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)

```

## Drawing oncoplots

for better visualization we are using oncoplot for top 10 mutated genes.

```{r, echo=TRUE}

#oncoplot for top ten mutated genes.
oncoplot(maf = laml, top = 10)

```

## Oncostrip

Visualizing specific set of genes using oncostrip and along with set of key words.

```{r, echo=TRUE}

oncostrip(maf = laml, genes = c('DNMT3A','NPM1', 'RUNX1'))

# Transition and Transversions

# Using titv() function to classify Snps into two stages transition or transversion which returns list of summarized tables which we can visualize using box plots.

laml.titv = titv(maf = laml, plot = FALSE, useSyn = TRUE)

# plot titv summary
plotTiTv(res = laml.titv)

# Lollipop plots for amino acid changes

# Plotting lollipop plots which is the most effective way of showing mutation spots on protein structures. These spots are usaully considered as mutational hot-spts and lollipop plots can be used to display rest of the mutations.

#lollipop plot for DNMT3A, which is one of the most frequent mutated gene in Leukemia.

lollipopPlot(maf = laml, gene = 'DNMT3A', AACol = 'Protein_Change', showMutationRate = TRUE)

#  Labelling points

# Labelling points on the lolliplot using aregument lablePos. To highlight the points in plot

lollipopPlot(maf = laml, gene = 'KIT', AACol = 'Protein_Change', labelPos = 816, refSeqID = 'NM_000222')

# Rainfall plots

# Plotting the rainfall plots as cancer genomes, especially solid tumors are located using genomic loci with hyper-mutations. Rainfall Plot displays highlight regions where potential changes in inter-event distances are located.

brca <- system.file("extdata", "brca.maf.gz", package = "maftools")
brca = read.maf(maf = brca, verbose = FALSE)

# Rainfall plot with detection points TRUE
#rainfallPlot(maf = brca, detectChangePoints = TRUE, pointSize = 0.6)

# Compare mutation load against TCGA cohorts

# TCGA contain 30 different cancers cohorts and median mutations, we see how mutations load in given maf and we use tcgacompare() function for that.

laml.mutload = tcgaCompare(maf = laml, cohortName = 'Example-LAML')

# Plotting VAF

# Fucntion calling fro variant allele frequencies

plotVaf(maf = laml, vafCol = 'i_TumorVAF_WU')

# Genecloud

# plot word cloud for mutated genes with function geneCLoud.

geneCloud(input = laml, minMut = 3)


```


## Processing copy-number data

# Reading and summarizing output files

```{r, echo =TRUE}

# Reading and summarizing gistic output files.

# Summarize output generated files by GISTIC programme

all.lesions <- system.file("extdata", "all_lesions.conf_99.txt", package = "maftools")
amp.genes <- system.file("extdata", "amp_genes.conf_99.txt", package = "maftools")
del.genes <- system.file("extdata", "del_genes.conf_99.txt", package = "maftools")
scores.gis <- system.file("extdata", "scores.gistic", package = "maftools")

laml.gistic = readGistic(gisticAllLesionsFile = all.lesions, gisticAmpGenesFile = amp.genes, gisticDelGenesFile = del.genes, gisticScoresFile = scores.gis, isTCGA = TRUE)

#GISTIC object
laml.gistic

# Visualizing gistic results

gisticChromPlot(gistic = laml.gistic, markBands = "all")

# Bubble plot

gisticBubblePlot(gistic = laml.gistic)

# oncoplot

gisticOncoPlot(gistic = laml.gistic, clinicalData = getClinicalData(x = laml), clinicalFeatures = 'FAB_classification', sortByAnnotation = TRUE, top = 10)

# Visualizing CBS segments

tcga.ab.009.seg <- system.file("extdata", "TCGA.AB.3009.hg19.seg.txt", package = "maftools")
plotCBSsegments(cbsFile = tcga.ab.009.seg)

```


## Analysis

```{r, echo=TRUE}

# Somatic Interactions

#exclusive/co-occurance event analysis on top 10 mutated genes. 
somaticInteractions(maf = laml, top = 25, pvalue = c(0.05, 0.1))

# Detecting cancer driver genes based on positional clustering

laml.sig = oncodrive(maf = laml, AACol = 'Protein_Change', minMut = 5, pvalMethod = 'zscore')
head(laml.sig)

plotOncodrive(res = laml.sig, fdrCutOff = 0.1, useFraction = TRUE)

# Adding and summarizing pfam domains

laml.pfam = pfamDomains(maf = laml, AACol = 'Protein_Change', top = 10)

#Protein summary (Printing first 7 columns for display convenience)
laml.pfam$proteinSummary[,1:7, with = FALSE]

#Domain summary (Printing first 3 columns for display convenience)
laml.pfam$domainSummary[,1:3, with = FALSE]

# Pan-Cancer comparison

#MutsigCV results for TCGA-AML
laml.mutsig <- system.file("extdata", "LAML_sig_genes.txt.gz", package = "maftools")
pancanComparison(mutsigResults = laml.mutsig, qval = 0.1, cohortName = 'LAML', inputSampleSize = 200, label = 1)

# Survival analysis

#Survival analysis based on grouping of DNMT3A mutation status
mafSurvival(maf = laml, genes = 'DNMT3A', time = 'days_to_last_followup', Status = 'Overall_Survival_Status', isTCGA = TRUE)

# Predict genesets associated with survival

#Using top 20 mutated genes to identify a set of genes (of size 2) to predict poor prognostic groups
prog_geneset = survGroup(maf = laml, top = 20, geneSetSize = 2, time = "days_to_last_followup", Status = "Overall_Survival_Status", verbose = FALSE)

print(prog_geneset)

mafSurvGroup(maf = laml, geneSet = c("DNMT3A", "FLT3"), time = "days_to_last_followup", Status = "Overall_Survival_Status")

# Comparing two cohorts (MAFs)

#Primary APL MAF
primary.apl = system.file("extdata", "APL_primary.maf.gz", package = "maftools")
primary.apl = read.maf(maf = primary.apl)
#Relapse APL MAF
relapse.apl = system.file("extdata", "APL_relapse.maf.gz", package = "maftools")
relapse.apl = read.maf(maf = relapse.apl)

#Considering only genes which are mutated in at-least in 5 samples in one of the cohort to avoid bias due to genes mutated in single sample.
pt.vs.rt <- mafCompare(m1 = primary.apl, m2 = relapse.apl, m1Name = 'Primary', m2Name = 'Relapse', minMut = 5)
print(pt.vs.rt)

# Forest plots

forestPlot(mafCompareRes = pt.vs.rt, pVal = 0.1, color = c('royalblue', 'maroon'), geneFontSize = 0.8)


# Co-onco plots

genes = c("PML", "RARA", "RUNX1", "ARID1B", "FLT3")
coOncoplot(m1 = primary.apl, m2 = relapse.apl, m1Name = 'PrimaryAPL', m2Name = 'RelapseAPL', genes = genes, removeNonMutated = TRUE)

# Lollipop plot-2

lollipopPlot2(m1 = primary.apl, m2 = relapse.apl, gene = "PML", AACol1 = "amino_acid_change", AACol2 = "amino_acid_change", m1_name = "Primary", m2_name = "Relapse")

# Clinical enrichment analysis

fab.ce = clinicalEnrichment(maf = laml, clinicalFeature = 'FAB_classification')

#Results are returned as a list. Significant associations p-value < 0.05
fab.ce$groupwise_comparision[p_value < 0.05]

plotEnrichmentResults(enrich_res = fab.ce, pVal = 0.05)


# Drug-Gene Interactions

dgi = drugInteractions(maf = laml, fontSize = 0.75)

dnmt3a.dgi = drugInteractions(genes = "DNMT3A", drugs = TRUE)

#Printing selected columns.
dnmt3a.dgi[,.(Gene, interaction_types, drug_name, drug_claim_name)]

# Oncogenic Signaling Pathways

OncogenicPathways(maf = laml)

PlotOncogenicPathways(maf = laml, pathways = "RTK-RAS")

#  Tumor heterogeneity and MATH scores

#Heterogeneity in sample TCGA.AB.2972

tcga.ab.2972.het = inferHeterogeneity(maf = laml, tsb = 'TCGA-AB-2972', vafCol = 'i_TumorVAF_WU')

print(tcga.ab.2972.het$clusterMeans)

#Visualizing results
plotClusters(clusters = tcga.ab.2972.het)

# Ignoring variants in copy number altered regions

seg = system.file('extdata', 'TCGA.AB.3009.hg19.seg.txt', package = 'maftools')
tcga.ab.3009.het = inferHeterogeneity(maf = laml, tsb = 'TCGA-AB-3009', segFile = seg, vafCol = 'i_TumorVAF_WU')

#Visualizing results. Highlighting those variants on copynumber altered variants.
plotClusters(clusters = tcga.ab.3009.het, genes = 'CN_altered', showCNvars = TRUE)

# Mutational Signatures

#Requires BSgenome object
library(BSgenome.Hsapiens.UCSC.hg19, quietly = TRUE)

laml.tnm = trinucleotideMatrix(maf = laml, prefix = 'chr', add = TRUE, ref_genome = "BSgenome.Hsapiens.UCSC.hg19")

# APOBEC Enrichment estimation.

plotApobecDiff(tnm = laml.tnm, maf = laml, pVal = 0.2)


# Signature analysis

library('NMF')
laml.sign = estimateSignatures(mat = laml.tnm, nTry = 6, pConstant = 1)

plotCophenetic(res = laml.sign)

laml.sig = extractSignatures(mat = laml.tnm, n = 3, pConstant = 1)

#Compate against original 30 signatures 
laml.og30.cosm = compareSignatures(nmfRes = laml.sig, sig_db = "legacy")

#Compate against updated version3 60 signatures 
laml.v3.cosm = compareSignatures(nmfRes = laml.sig, sig_db = "SBS")

library('pheatmap')
pheatmap::pheatmap(mat = laml.og30.cosm$cosine_similarities, cluster_rows = FALSE, main = "cosine similarity against validated signatures")

maftools::plotSignatures(nmfRes = laml.sig, title_size = 0.8)

# Signature enrichment analysis

#laml.se = signatureEnrichment(maf = laml, sig_res = laml.sig)

#plotEnrichmentResults(enrich_res = laml.se, pVal = 0.05)

```

